### PR TITLE
Fix sec vuln with list of claims

### PR DESCRIPTION
**Pull request by Waterdrips**
Fix for Security Vuln in v 3.2.0 with invalid check on "aud" claim when claim is a list of strings (RFC states it should be list of strings, or in special case a String only)

Signed-off-by: Alistair Hey <alistair@heyal.co.uk>

### CODE REVIEW

**Review by leo-form3 — DISMISSED** [review]

**Review by wojciecholszewski-form3 — DISMISSED** [review]

File: map_claims.go [hunk 1]
```diff
@@ -13,7 +13,14 @@ type MapClaims map[string]interface{}
 // Compares the aud claim against cmp.
 // If required is false, this method will return true if the value matches or is unset
 func (m MapClaims) VerifyAudience(cmp string, req bool) bool {
-	aud, _ := m["aud"].(string)
+	aud, _ := m["aud"].([]string)
```
Comment (rkorkosz) [inline map_claims.go]: I would make it explicit and use `ok` as second value, then you can check if it's `!ok` proceed as string.

File: map_claims.go [hunk 2]
```diff
@@ -13,7 +13,14 @@ type MapClaims map[string]interface{}
...
+
+	// special case where aud is a single string not list of strings
+	if aud == nil {
+		strAud, _ := m["aud"].(string)
```
Comment (rkorkosz) [inline map_claims.go]: Same here, use `ok` and check if it's actually a string.

**Review by rkorkosz — CHANGES_REQUESTED** [review]

File: map_claims.go (hunk 1 above)
Comment (Waterdrips) [reply map_claims.go]: if this first one fails we need to do the check on --> .(string)

Iv updated the section to match.

**Review by rkorkosz — APPROVED** [review]

File: example_test.go [hunk 3]
```diff
@@ -28,7 +28,7 @@ func ExampleNewWithClaims_standardClaims() {
 
 // Example creating a token using a custom claims type.  The StandardClaim is embedded
 // in the custom type to allow for easy encoding, parsing and validation of standard claims.
-func ExampleNewWithClaims_customClaimsType() {
+func NewWithClaims_customClaimsType() {
```
Comment (wojciecholszewski-form3) [inline example_test.go:31]: is it a desired rename?

File: example_test.go [hunk 4]
```diff
@@ -53,7 +53,7 @@ func ExampleNewWithClaims_customClaimsType() {
 
 // Example creating a token using a custom claims type.  The StandardClaim is embedded
 // in the custom type to allow for easy encoding, parsing and validation of standard claims.
-func ExampleParseWithClaims_customClaimsType() {
+func ParseWithClaims_customClaimsType() {
```
Comment (wojciecholszewski-form3) [inline example_test.go:56]: as above, is it desired change?

File: example_test.go (hunk 3 above)
Comment (wojciecholszewski-form3) [reply example_test.go:31]: discussed f2f

File: example_test.go (hunk 4 above)
Comment (wojciecholszewski-form3) [reply example_test.go:56]: discussed f2f

**Review by wojciecholszewski-form3 — DISMISSED** [review]

**Review by owenrumney — DISMISSED** [review]
lgtm

**Review by rkorkosz — DISMISSED** [review]

**Comment by leo-form3** [comment]
Travis is having issues building the branch. Is it maybe because you are creating a PR from your private repo?

**Review by owenrumney — APPROVED** [review]

**Review by wojciecholszewski-form3 — APPROVED** [review]

**Comment by Waterdrips** [comment]
close and re-open to try and get rid of duplicate travis PR checks

**Comment by Waterdrips** [comment]
going to re-open a new pr to try and fix the duplicate checks

**Pull request closed without merging**